haystack/agents/base.py

```diff
@@ -111,6 +111,120 @@ def _process_result(self, result: Any) -> str:
+                name="Calculator",
+                pipeline_or_node=calculator
+                description="Useful when you need to answer questions about math"
+            )
+        )
+        """
+        self.tools[tool.name] = tool
+
+    def get_tool_names(self) -> str:
+        """
+        Returns a string with the names of all tools registered
+        """
+        return ", ".join(self.tools.keys())
+
+    def get_tools(self) -> List[Tool]:
+        """
+        Returns a list of all tool instances registered
+        """
+        return list(self.tools.values())
+
+    def get_tool_names_with_descriptions(self) -> str:
+        """
+        Returns a string with the names and descriptions of all tools registered
+        """
+        return "\n".join([f"{tool.name}: {tool.description}" for tool in self.tools.values()])
+
+    def has_tools(self) -> bool:
+        """
+        Returns True if this ToolsManager has any tools, False otherwise.
+        """
+        return len(self.tools) > 0
+
+    def has_tool(self, tool_name: str) -> bool:
+        """
+        Returns True if this ToolsManager has a tool with the given name, False otherwise.
+
+        :param tool_name: The name of the tool for which you want to check whether the ToolsManager has it.
+        """
+        return tool_name in self.tools
+
+    def run_tool(self, llm_response: str, params: Optional[Dict[str, Any]] = None) -> str:
+        tool_result: str = ""
+        if self.has_tools():
+            tool_name, tool_input = self.extract_tool_name_and_tool_input(llm_response)
+            if tool_name and tool_input:
+                tool: Tool = self.tools[tool_name]
+                try:
+                    self.callback_manager.on_tool_start(tool_input, tool=tool)
+                    tool_result = tool.run(tool_input, params)
+                    self.callback_manager.on_tool_finish(
+                        tool_result,
+                        observation_prefix="Observation: ",
+                        llm_prefix="Thought: ",
+                        color=tool.logging_color,
+                    )
+                except Exception as e:
+                    self.callback_manager.on_tool_error(e, tool=self.tools[tool_name])
+                    raise e
+        return tool_result
+
+    def extract_tool_name_and_tool_input(self, llm_response: str) -> Tuple[Optional[str], Optional[str]]:
+        """
+        Parse the tool name and the tool input from the PromptNode response.
+        :param llm_response: The PromptNode response.
+        :return: A tuple containing the tool name and the tool input.
+        """
+        tool_match = re.search(self.tool_pattern, llm_response)
+        if tool_match:
+            tool_name = tool_match.group(1)
+            tool_input = tool_match.group(3)
+            return tool_name.strip('" []\n').strip(), tool_input.strip('" \n')
+        return None, None
+
+    def valid_tool_selected(self, llm_response: str) -> bool:
```

Comment:
Where is this used? The code is pretty much the same as `extract_tool_name_and_tool_input`, I would rather make the other one raise if the tool suggested is not valid and get rid of this.